reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java

```diff
@@ -578,9 +663,193 @@ public Object scanUnsafe(Attr key) {
 
 		@Override
 		public boolean isDisposed() {
-			return s == Operators.cancelledSubscription() || done;
+			long state = this.state;
+			return isTerminated(state) || isCancelled(state);
 		}
 
+		static void clearAndFinalize(PublishSubscriber<?> instance) {
+			for (; ; ) {
+				final long state = instance.state;
+
+				if (isFinalized(state)) {
+					instance.clear();
+					return;
+				}
+
+				if (isSubscriptionSet(state)) {
+					instance.clear();
+				}
+
+				if (STATE.compareAndSet(
+						instance, state,
+						(state & ~WORK_IN_PROGRESS_MASK) | FINALIZED_FLAG)) {
+					break;
+				}
+			}
+		}
+
+		static long addWork(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (STATE.compareAndSet(instance, state, addWork(state))) {
+					return state;
+				}
+			}
+		}
+
+		static long addWorkIfSubscribed(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (!isSubscriptionSet(state)) {
+					return state;
+				}
+
+				if (STATE.compareAndSet(instance, state, addWork(state))) {
+					return state;
+				}
+			}
+		}
+
+		static long addWork(long state) {
+			if ((state & WORK_IN_PROGRESS_MASK) == WORK_IN_PROGRESS_MASK) {
+				return (state &~ WORK_IN_PROGRESS_MASK) | 1;
+			}
+			else {
+				return state + 1;
```

Comment:
do we need to guard against overflow with WORK_IN_PROGRESS_MASK?

Reply:
it is guardede on the line 716 (if state & WIPM == WIPM it means that all mask cells are 1 which means max value is reached so the branch on line 716 is called and max WIP values is repalces with 1

Reply:
Understood, thanks!